rcon/commands.py

```diff
@@ -73,78 +89,119 @@ class ServerCtl:
     set password not implemented on purpose
     """
 
-    def __init__(self, config, auto_retry=1):
+    def __init__(self, config, auto_retry=1, pool_size=20):
         # .env fed config from rcon.SERVER_INFO
         self.config = config
-        self.conn = None
-        # self._connect()
         self.auto_retry = auto_retry
+        self.mu = threading.RLock()
+        self.idles: list[HLLConnection] = []
+        self.numOpen = 0
+        self.maxOpen = 20
+        self.maxIdle = pool_size
+
+    @contextmanager
+    def with_connection(self) -> HLLConnection:
+        logger.debug("Waiting to acquire lock")
+        with self.mu:
+            if len(self.idles) != 0:
+                logger.debug("acquiring connection from idle pool")
+                conn = self.idles.pop()
+
+            elif self.numOpen >= self.maxOpen:
+                logger.debug("Max connections already open, waiting for connection returned to pool")
+                c = 0
+                while len(self.idles) == 0:
+                    if c >= 30:
+                        logger.error("waiting for connection returned to pool timed out after " + str(c) + " seconds")
+                        raise TimeoutError()
+                    c += 1
+                    time.sleep(1)
+                conn = self.idles.pop()
+
+            else:
+                logger.debug("Opening a new connection")
+                conn = HLLConnection()
+                self._connect(conn)
+                self.numOpen += 1
 
-    def _connect(self):
-        self.conn = HLLConnection()
         try:
-            self.conn.connect(
-                self.config["host"], int(self.config["port"]), self.config["password"]
-            )
+            yield conn
+        finally:
+            logger.debug("return connection")
+            if len(self.idles) >= self.maxIdle:
+                logger.debug("Enough connections in pool, closing")
+                self.numOpen -= 1
+                conn.close()
+            else:
+                logger.debug("Returning connection to pool")
+                self.idles.append(conn)
```

Comment:
hmm I'm not sure it's the best move to add the connection in the "finally" cause what if that connection just raised cause it was a bad password, or just somehow it got disconnected. We could shove a broken connection in the pool

Reply:
I was thinking about his as well.

For the wrong password case, this should already be handled when the connection is setup in the _connect method. So this error should raise all the way up to the caller.
For the errors that can happen when an action si executed: Most of the commands have a handling of that already (and either call reconnect or do some other handling). I'm not sure, if we should change that in this commit, tbh. Right now, if a connection is broken, we keep using it in the same thread as well, with the same consequences.
Overhauling error handling of such basic RCon errors should probably be done in a separate PR? Or what do you think? :)